In [1]:
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:90% !important; }</style>"))
from IPython.core.interactiveshell import InteractiveShell  #to run all statements in cell, not only the last
InteractiveShell.ast_node_interactivity = "all"

import warnings
warnings.filterwarnings('ignore')

from sklearn.svm import SVC, NuSVC, LinearSVC, OneClassSVM, SVR, NuSVR, LinearSVR
from sklearn.linear_model import Ridge
from typing import Dict, Tuple
from IPython.display import Image
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris, load_boston
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.model_selection import cross_val_score, cross_validate
from sklearn.model_selection import KFold, RepeatedKFold, LeaveOneOut, LeavePOut, ShuffleSplit, StratifiedKFold
from sklearn.metrics import accuracy_score, balanced_accuracy_score
from sklearn.metrics import precision_score, recall_score, f1_score, classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import mean_absolute_error, mean_squared_error, mean_squared_log_error, median_absolute_error, r2_score 
from sklearn.metrics import roc_curve, roc_auc_score
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.model_selection import learning_curve, validation_curve
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline 
sns.set(style="ticks")

from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.ensemble import ExtraTreesClassifier, ExtraTreesRegressor
from sklearn.ensemble import GradientBoostingClassifier, GradientBoostingRegressor
from sklearn.ensemble import BaggingClassifier
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor, export_graphviz
from sklearn.linear_model import LogisticRegression
from xgboost import XGBClassifier
from sklearn.metrics import average_precision_score
from sklearn.calibration import CalibratedClassifierCV

In [2]:
data = pd.read_csv('heart.csv', sep=",")
data.head(5)
data.dtypes

features = data.drop(['target'], axis=1)
target_tmp = data['target']

target = pd.DataFrame({'target':target_tmp.index, 'target':target_tmp.values})

features.head(5)
type(features)
features.shape
target.head(5)
type(target)
target.shape

UCI_X_train, UCI_X_test, UCI_y_train, UCI_y_test = train_test_split(features, target, test_size=0.1, random_state=1)
UCI_X_train.shape
UCI_X_test.shape
UCI_y_train.shape
UCI_y_test.shape

type(UCI_X_train)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
thal          int64
target        int64
dtype: object

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

pandas.core.frame.DataFrame

(303, 13)

target
0       1
1       1
2       1
3       1
4       1

pandas.core.frame.DataFrame

(303, 1)

(272, 13)

(31, 13)

(272, 1)

(31, 1)

pandas.core.frame.DataFrame

### Random forest

In [3]:
cl1 = RandomForestClassifier(random_state=1).fit(UCI_X_train, UCI_y_train)

pred_iris_y_test = cl1.predict(UCI_X_test)
pred_iris_y_test
UCI_y_test['target'].values

array([0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0,
       0, 0, 1, 1, 0, 1, 1, 1, 1], dtype=int64)

array([0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 1], dtype=int64)

In [4]:
accuracy_score(UCI_y_test, pred_iris_y_test)
precision_score(UCI_y_test, pred_iris_y_test), recall_score(UCI_y_test, pred_iris_y_test)
f1_score(UCI_y_test, pred_iris_y_test)

0.8387096774193549

(0.8125, 0.8666666666666667)

0.8387096774193549

In [5]:
scores_loo = cross_val_score(RandomForestClassifier(random_state=1), 
                         features, target, 
                         cv=LeaveOneOut())
np.mean(scores_loo)

0.801980198019802

### Gradient boosting

In [6]:
cl1 = GradientBoostingClassifier(random_state=1).fit(UCI_X_train, UCI_y_train)

pred_iris_y_test = cl1.predict(UCI_X_test)
pred_iris_y_test
UCI_y_test['target'].values

array([0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 1], dtype=int64)

array([0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 1], dtype=int64)

In [7]:
accuracy_score(UCI_y_test, pred_iris_y_test)
precision_score(UCI_y_test, pred_iris_y_test), recall_score(UCI_y_test, pred_iris_y_test)
f1_score(UCI_y_test, pred_iris_y_test)

0.8709677419354839

(0.8235294117647058, 0.9333333333333333)

0.8749999999999999

In [8]:
scores_loo = cross_val_score(GradientBoostingClassifier(random_state=1), 
                         features, target, 
                         cv=KFold(n_splits=10))
np.mean(scores_loo)

0.7684946236559139

#### Some grid search for grad boost

In [9]:
n_range = np.array([x / 100.0 for x in range(1, 20, 1)])
tuned_parameters = [{'learning_rate': n_range}]
tuned_parameters

[{'learning_rate': array([0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 , 0.11,
         0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19])}]

In [10]:
clf_gs = GridSearchCV(GradientBoostingClassifier(n_estimators=2), tuned_parameters, cv=8, scoring='f1', verbose=1, n_jobs=-1)
clf_gs.fit(UCI_X_train, UCI_y_train)

Fitting 8 folds for each of 19 candidates, totalling 152 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  42 tasks      | elapsed:   12.3s
[Parallel(n_jobs=-1)]: Done 152 out of 152 | elapsed:   16.3s finished
C:\Users\bulat\AppData\Roaming\Python\Python36\site-packages\sklearn\model_selection\_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


GridSearchCV(cv=8, error_score='raise-deprecating',
       estimator=GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.1, loss='deviance', max_depth=3,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_sampl...      subsample=1.0, tol=0.0001, validation_fraction=0.1,
              verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=-1,
       param_grid=[{'learning_rate': array([0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 , 0.11,
       0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19])}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='f1', verbose=1)

In [11]:
clf_gs.cv_results_

{'mean_fit_time': array([0.15646115, 0.1834529 , 0.11047724, 0.09497651, 0.15895924,
        0.23244259, 0.12047014, 0.16445974, 0.0864791 , 0.14196619,
        0.13046733, 0.16845867, 0.09747186, 0.13096517, 0.10297269,
        0.13146472, 0.05798542, 0.12646854, 0.13796413]),
 'mean_score_time': array([0.0059976 , 0.00299919, 0.00449562, 0.00449941, 0.0039992 ,
        0.00449777, 0.00399858, 0.00400004, 0.00400162, 0.00449669,
        0.00349981, 0.00449789, 0.00349969, 0.00299889, 0.00350013,
        0.00450051, 0.00399962, 0.00249869, 0.003001  ]),
 'mean_test_score': array([0.71087106, 0.71087106, 0.71087106, 0.71087106, 0.75143179,
        0.80313848, 0.80633264, 0.81722788, 0.8419614 , 0.85216738,
        0.85216738, 0.85215687, 0.85215687, 0.84878872, 0.84617548,
        0.84617548, 0.84349524, 0.84349524, 0.84610848]),
 'mean_train_score': array([0.71090025, 0.71090025, 0.71090025, 0.71090025, 0.77106565,
        0.82354159, 0.83648493, 0.8512899 , 0.8660246 , 0.87065012,
   

In [12]:
clf_gs.best_estimator_
clf_gs.best_score_
clf_gs.best_params_

GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.1, loss='deviance', max_depth=3,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=2,
              n_iter_no_change=None, presort='auto', random_state=None,
              subsample=1.0, tol=0.0001, validation_fraction=0.1,
              verbose=0, warm_start=False)

0.8521673792241591

{'learning_rate': 0.1}

### Stacking

In [13]:
from heamy.estimator import Regressor, Classifier
from heamy.pipeline import ModelsPipeline
from heamy.dataset import Dataset

In [14]:
def val_mae(model):
    model.fit(UCI_X_train, UCI_y_train)
    pred_iris_y_test = model.predict(UCI_X_test)
    result = average_precision_score(UCI_y_test, pred_iris_y_test)
    print(model)
    print('average_precision_score={}'.format(result))

In [15]:
val_mae(LogisticRegression())
val_mae(SVC())
val_mae(XGBClassifier())
val_mae(RandomForestClassifier())

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)
average_precision_score=0.8008855154965211
SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='rbf', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)
average_precision_score=0.4838709677419355
XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1)
average_precision_score=0.

In [16]:
# набор данных
dataset = Dataset(UCI_X_train, UCI_y_train, UCI_X_test)

# модели первого уровня
model_1 = Classifier(dataset=dataset, estimator=LogisticRegression, name='log_reg')
model_2 = Classifier(dataset=dataset, estimator=SVC, parameters={'probability': True}, name='svm')
model_3 = Classifier(dataset=dataset, estimator=XGBClassifier,name='xgb')
model_4 = Classifier(dataset=dataset, estimator=RandomForestClassifier,name='rf')

In [17]:
# Stacking трех моделей
pipeline = ModelsPipeline(model_1, model_2, model_3, model_4)
stack_ds = pipeline.stack(k=10, seed=1)

# модель второго уровня
stacker = Classifier(dataset=stack_ds, estimator=LogisticRegression)
results = stacker.validate(k=10,scorer=average_precision_score)

Metric: average_precision_score
Folds accuracy: [0.9846415489272631, 0.8739283978835064, 0.9595394801277154, 0.8949456101697317, 0.9626696832579186, 0.9742002063983488, 0.8532975058710354, 0.88369559561815, 0.9723622782446312, 0.8359925788497217]
Mean accuracy: 0.9195272885348021
Standard Deviation: 0.05367926273835042
Variance: 0.002881463248132856


In [18]:
# Stacking трех моделей
pipeline = ModelsPipeline(model_1, model_4)
stack_ds = pipeline.stack(k=10, seed=1)

# модель второго уровня
stacker = Classifier(dataset=stack_ds, estimator=LogisticRegression)
results = stacker.validate(k=10,scorer=average_precision_score)

Metric: average_precision_score
Folds accuracy: [0.9846415489272631, 0.8679760169311255, 0.952132072720308, 0.8984158241662614, 0.9626696832579186, 0.9859649122807017, 0.8592190016103062, 0.8960885468697969, 0.9757936507936508, 0.8343259121830551]
Mean accuracy: 0.9217227169740386
Standard Deviation: 0.054078492525025
Variance: 0.002924483353779185


### Result by Precision

Random forest: 0.8125
Gradient boosting: 0.8235  
Stacking: 0.9217  
    
BEST Stacking